pkg/apis/elasticsearch/v1/autoscaling.go

```diff
@@ -0,0 +1,243 @@
+// you may not use this file except in compliance with the Elastic License.
+
+package v1
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common/version"
+	"github.com/elastic/cloud-on-k8s/pkg/utils/set"
+	"k8s.io/apimachinery/pkg/api/resource"
+)
+
+const ElasticsearchAutoscalingSpecAnnotationName = "elasticsearch.alpha.elastic.co/autoscaling-spec"
+
+var errNodeRolesNotSet = errors.New("node.roles must be set")
+
+// -- Elasticsearch Autoscaling API structures
+
+// DeciderSettings allows the user to tweak autoscaling deciders.
+// The map data structure complies with the <key,value> format expected by Elasticsearch.
+// +kubebuilder:object:generate=false
+type DeciderSettings map[string]string
+
+// AutoscalingPolicy models the Elasticsearch autoscaling API.
+// +kubebuilder:object:generate=false
+type AutoscalingPolicy struct {
+	// An autoscaling policy must target a unique set of roles.
+	Roles []string `json:"roles,omitempty"`
+	// Deciders allows the user to override default settings for autoscaling deciders.
+	Deciders map[string]DeciderSettings `json:"deciders,omitempty"`
+}
+
+// -- Elastic Cloud on K8S specific structures
+
+// AutoscalingSpec is the root object of the autoscaling specification in the Elasticsearch resource definition.
+// +kubebuilder:object:generate=false
+type AutoscalingSpec struct {
+	AutoscalingPolicySpecs AutoscalingPolicySpecs `json:"policies"`
+	// Elasticsearch is stored in the autoscaling spec for convenience. It should be removed once the autoscaling spec is
+	// fully part of the Elasticsearch specification.
+	Elasticsearch Elasticsearch `json:"-"`
+}
+
+// +kubebuilder:object:generate=false
+type AutoscalingPolicySpecs []AutoscalingPolicySpec
+
+// NamedAutoscalingPolicy models an autoscaling policy as expected by the Elasticsearch policy API.
+// It is identified by a unique name provided by the user.
+// +kubebuilder:object:generate=false
+type NamedAutoscalingPolicy struct {
+	// Name identifies the autoscaling policy in the autoscaling specification.
+	Name string `json:"name,omitempty"`
+	// AutoscalingPolicy is the autoscaling policy as expected by the Elasticsearch API.
+	AutoscalingPolicy
+}
+
+// AutoscalingPolicySpec holds a named autoscaling policy and the associated resources limits (cpu, memory, storage).
+// +kubebuilder:object:generate=false
+type AutoscalingPolicySpec struct {
+	NamedAutoscalingPolicy
+
+	AutoscalingResources `json:"resources"`
+}
+
+// +kubebuilder:object:generate=false
+// AutoscalingResources models the limits, submitted by the user, for the supported resources in an autoscaling policy.
+// Only the node count range is mandatory. For other resources, a limit range is required only
+// if the Elasticsearch autoscaling capacity API returns a requirement for a given resource.
+// For example, the memory limit range is only required if the autoscaling API response contains a memory requirement.
```

Comment:
It is not possible to know in advance the required resources for a role, it can only be detected after the call to the capacity endpoint.
What will happen in such a case:
* The node sets managed by the autoscaling policy will not be reconciled.
* The situation will be surfaced in the status and in the logs (we should also surface it through some warning events)